From ae991ba20a1e227ee9cadfc5631a6ac007f73f4d Mon Sep 17 00:00:00 2001
From: Yann Gautier <yann.gautier@foss.st.com>
Date: Wed, 15 Mar 2023 15:11:05 +0100
Subject: [PATCH 307/954] feat(stm32mp2): add a low-power context

Add a low-power context struct, and a dedicated file to manage it
(plat/st/stm32mp2/stm32mp2_context.c). This structure holds the key used
to encrypt some DDR areas, and other low-power contexts. New helper
functions are added to save and get those variables for low-power.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git a15d31168b8f9cabec2548e842765a7befd1b2e8 feat(stm32mp2): add a low-power context]
Signed-off-by: Yann Gautier <yann.gautier@foss.st.com>
Change-Id: I05996008b108ca67f280af5f7cadc4de5e244382
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/295672
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 plat/st/stm32mp2/include/stm32mp2_context.h |  21 +++
 plat/st/stm32mp2/platform.mk                |   1 +
 plat/st/stm32mp2/stm32mp2_context.c         | 158 ++++++++++++++++++++
 3 files changed, 180 insertions(+)
 create mode 100644 plat/st/stm32mp2/include/stm32mp2_context.h
 create mode 100644 plat/st/stm32mp2/stm32mp2_context.c

diff --git a/plat/st/stm32mp2/include/stm32mp2_context.h b/plat/st/stm32mp2/include/stm32mp2_context.h
new file mode 100644
index 0000000000..bb18b49b3e
--- /dev/null
+++ b/plat/st/stm32mp2/include/stm32mp2_context.h
@@ -0,0 +1,21 @@
+/*
+ * Copyright (c) 2023, ARM Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef STM32MP2_CONTEXT_H
+#define STM32MP2_CONTEXT_H
+
+#include <stdint.h>
+
+#include <lib/psci/psci.h>
+
+void stm32mp_pm_save_enc_mkey_in_context(uint8_t *data);
+void stm32mp_pm_get_enc_mkey_from_context(uint8_t *data);
+
+bool stm32_pm_context_is_valid(void);
+void stm32_pm_context_save(const psci_power_state_t *state);
+void stm32_pm_context_restore(void);
+
+#endif /* STM32MP2_CONTEXT_H */
diff --git a/plat/st/stm32mp2/platform.mk b/plat/st/stm32mp2/platform.mk
index 5851180a8c..efbc1d4157 100644
--- a/plat/st/stm32mp2/platform.mk
+++ b/plat/st/stm32mp2/platform.mk
@@ -183,6 +183,7 @@ PLAT_BL_COMMON_SOURCES		+=	plat/st/stm32mp2/stm32mp2_private.c
 PLAT_BL_COMMON_SOURCES		+=	drivers/st/bsec/bsec3.c					\
 					drivers/st/iwdg/stm32_iwdg.c				\
 					drivers/st/reset/stm32mp2_reset.c			\
+					plat/st/stm32mp2/stm32mp2_context.c			\
 					plat/st/stm32mp2/stm32mp2_pwr.c				\
 					plat/st/stm32mp2/stm32mp2_syscfg.c
 
diff --git a/plat/st/stm32mp2/stm32mp2_context.c b/plat/st/stm32mp2/stm32mp2_context.c
new file mode 100644
index 0000000000..4317179622
--- /dev/null
+++ b/plat/st/stm32mp2/stm32mp2_context.c
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2023, ARM Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <drivers/clk.h>
+#include <lib/el3_runtime/context_mgmt.h>
+#include <lib/el3_runtime/cpu_data.h>
+
+#if defined(SPD_opteed)
+#include "../../../services/spd/opteed/opteed_private.h"
+#endif
+
+#include "../../../lib/psci/psci_private.h"
+
+#include <platform_def.h>
+#include <stm32mp2_context.h>
+
+#define BACKUP_CTX_ADDR		STM32MP_BACKUP_RAM_BASE
+#define BACKUP_CTX_CLK		CK_BUS_BKPSRAM
+#define ENC_KEY_SIZE_IN_BYTES	RISAF_KEY_SIZE_IN_BYTES
+
+/* Magic used to indicated valid = ' ' 'M' 'P' '2' */
+#define CONTEXT_MAGIC			0x204D5032
+
+struct backup_data_s {
+	uint32_t magic;
+	uint8_t enc_mkey[ENC_KEY_SIZE_IN_BYTES];
+	psci_power_state_t standby_pwr_state;
+	cpu_context_t saved_cpu_s_context[PLATFORM_CORE_COUNT];
+	cpu_context_t saved_cpu_ns_context[PLATFORM_CORE_COUNT];
+	suspend_mode_t psci_suspend_mode;
+#if defined(SPD_opteed)
+	uintptr_t optee_vector;
+	optee_context_t opteed_sp_context[OPTEED_CORE_COUNT];
+#endif
+};
+
+void stm32mp_pm_save_enc_mkey_in_context(uint8_t *data)
+{
+	struct backup_data_s *backup_data;
+
+	backup_data = (struct backup_data_s *)BACKUP_CTX_ADDR;
+
+	clk_enable(BACKUP_CTX_CLK);
+
+	memcpy(backup_data->enc_mkey, data, ENC_KEY_SIZE_IN_BYTES);
+
+	clk_disable(BACKUP_CTX_CLK);
+}
+
+void stm32mp_pm_get_enc_mkey_from_context(uint8_t *data)
+{
+	struct backup_data_s *backup_data;
+
+	backup_data = (struct backup_data_s *)BACKUP_CTX_ADDR;
+
+	clk_enable(BACKUP_CTX_CLK);
+
+	memcpy(data, backup_data->enc_mkey, ENC_KEY_SIZE_IN_BYTES);
+
+	clk_disable(BACKUP_CTX_CLK);
+}
+
+bool stm32_pm_context_is_valid(void)
+{
+	struct backup_data_s *backup_data;
+	bool ret;
+
+	backup_data = (struct backup_data_s *)BACKUP_CTX_ADDR;
+
+	clk_enable(BACKUP_CTX_CLK);
+
+	ret = (backup_data->magic == CONTEXT_MAGIC);
+	clk_disable(BACKUP_CTX_CLK);
+
+	return ret;
+}
+
+void stm32_pm_context_save(const psci_power_state_t *state)
+{
+	void *cpu_context;
+	struct backup_data_s *backup_data;
+
+	clk_enable(BACKUP_CTX_CLK);
+	backup_data = (struct backup_data_s *)BACKUP_CTX_ADDR;
+
+	backup_data->magic = CONTEXT_MAGIC;
+
+	/* Retrieve non-secure CPU context struct address */
+	cpu_context = cm_get_context(NON_SECURE);
+
+	/* Save context in Backup SRAM */
+	memcpy(&backup_data->saved_cpu_ns_context[0], cpu_context,
+	       sizeof(cpu_context_t) * PLATFORM_CORE_COUNT);
+
+	/* Retrieve secure CPU context struct address */
+	cpu_context = cm_get_context(SECURE);
+
+	/* Save context in Backup SRAM */
+	memcpy(&backup_data->saved_cpu_s_context[0], cpu_context,
+	       sizeof(cpu_context_t) * PLATFORM_CORE_COUNT);
+
+#if defined(SPD_opteed)
+	backup_data->optee_vector = (uintptr_t)optee_vector_table;
+
+	memcpy(&backup_data->opteed_sp_context[0], opteed_sp_context,
+	       sizeof(optee_context_t) * OPTEED_CORE_COUNT);
+#endif
+
+	/* Save PSCI state in Backup SRAM */
+	memcpy(&backup_data->standby_pwr_state, state, sizeof(psci_power_state_t));
+	backup_data->psci_suspend_mode = psci_suspend_mode;
+
+	clk_disable(BACKUP_CTX_CLK);
+}
+
+void stm32_pm_context_restore(void)
+{
+	void *cpu_context;
+	struct backup_data_s *backup_data;
+
+	clk_enable(BACKUP_CTX_CLK);
+	backup_data = (struct backup_data_s *)BACKUP_CTX_ADDR;
+
+	/* Retrieve non-secure CPU context struct address */
+	cpu_context = cm_get_context(NON_SECURE);
+
+	/* Restore data from Backup SRAM */
+	memcpy(cpu_context, backup_data->saved_cpu_ns_context,
+	       sizeof(cpu_context_t) * PLATFORM_CORE_COUNT);
+
+	/* Retrieve non-secure CPU context struct address */
+	cpu_context = cm_get_context(SECURE);
+
+	/* Restore data from Backup SRAM */
+	memcpy(cpu_context, backup_data->saved_cpu_s_context,
+	       sizeof(cpu_context_t) * PLATFORM_CORE_COUNT);
+
+	psci_set_target_local_pwr_states(PLAT_MAX_PWR_LVL,
+					 &backup_data->standby_pwr_state);
+
+	if (psci_set_suspend_mode(backup_data->psci_suspend_mode) != PSCI_E_SUCCESS) {
+		panic();
+	}
+
+#if defined(SPD_opteed)
+	optee_vector_table = (optee_vectors_t *)backup_data->optee_vector;
+
+	memcpy(opteed_sp_context, backup_data->opteed_sp_context,
+	       sizeof(optee_context_t) * OPTEED_CORE_COUNT);
+
+	opteed_restore();
+#endif
+
+	clk_disable(BACKUP_CTX_CLK);
+}
-- 
2.53.0

